p4-16/psa/PSA.mdk

```diff
@@ -309,6 +310,134 @@ Additional supported match_kind types
++ Fields in the data contents of a `Register` array (section
+  [#sec-registers]).
++ Match fields of an entry in a `ValueSet` extern instance (section
+  [#sec-parser-value-sets]).
+
+Note: There may be other channels not listed above.
+
+For packets between the control plane and the PSA device, there is the
+issue that many PSA implementations are expected to restrict P4
+programs to have headers with fields with a total length that is a
+multiple of 8 bits.  To make it possible to define P4 header types
+that meet this restriction, and contain values of fields with these
+PSA-specific types, and be source-compatible across multiple PSA
+implementations, additional types are defined that contain `InHeader` in
+their name.  For example, `PortIdInHeader_t` is the same as `PortId_t`,
+except it must be a multiple of 8 bits long, and contain at least as
+many bits as `PortId_t` does.
+
+Because these types are guaranteed to
+be a multiple of 8 bits long, any combination of them may be included
+in a P4 header type definition, as long as the other fields in the
+header satisfy the multiple of 8 bits restriction.  The controller or
+P4 program generating packets with such headers should fill in any
+most significant "padding" bits with 0.  In a P4 program, this can be
+done with a normal assignment statement, where the value on the right
+hand side is cast to the wider `InHeader` type.
+
+Values of type `PortId_t` have a distinctive property in PSA
+implementations.  Because it can make some hardware implementations
+more straightforward, the numerical values of fields with type
+`PortId_t` in the P4 data plane might not be a simple range of values
+such as 0 through 31, as one might choose when writing control plane
+software for a 32-port device.  The agent is expected to implement
+numerical translation between controller port id values and data plane
+port id values, for each of the channels of communication between the
+controller and data plane described above.
+
+Occurrences of values of type `PortId_t` or `PortIdInHeader_t` must
+include a `port_translation` annotation so the compiler can mark
+occurrences of these values appropriately, so the agent can know which
+values need this translation.  See the example P4 header definition
+below.
+
+```
+[INCLUDE=examples/psa-example-parser-error-handling.p4:PortId_Annotation_Example]
+```
+
+Because of this translation, it is recommended that values of type
+`PortId_t` be treated similarly to values of an `enum` type.
+Comparing two values of this type for equality or inequality is
+reasonable, as well as assigning the values to other variables or
+parameters of the same type, but nearly any other operations are prone
+to error.  When matching values of type `PortId_t` as part of a table
+key, always match a complete value exactly, or wildcard every bit of
+the value (i.e. a `ternary` match kind with all bits wildcard, or an
+`lpm` match kind with prefix length 0).  If you attempt to do any of
```

Comment:
I believe P4Runtime disallows such behavior: all wildcards or LPM with prefix of 0. @antoninbas can confirm.

Reply:
@cc10512 I am pretty sure P4Runtime allows one to create table entries that completely wildcard ternary fields, or do length 0 prefix of lpm fields -- it is just that they represent it in a way where the field isn't even mentioned in the P4Runtime API message.

Reply:
@jafingerhut is correct; we do allow wildcards for ternary & LPM but require a certain formatting for the messages.